File: King_s_Order.py
class DSU:
    def __init__(self, n):
        self.parent = [-1] * (n + 1)
        self.parentLevel = [0] * (n + 1)
        self.parentSize = [1] * (n + 1)

    def find(self, n):
        if self.parent[n] == -1:
            return n
        leader = self.find(self.parent[n])
        self.parent[n] = leader
        return leader

    def union(self, a, b):
        leaderA = self.find(a)
        leaderB = self.find(b)
        if leaderA != leaderB:
            if self.parentLevel[leaderA] > self.parentLevel[leaderB]:
                self.parent[leaderB] = leaderA
            elif self.parentLevel[leaderB] > self.parentLevel[leaderA]:
                self.parent[leaderA] = leaderB
            else:
                self.parent[leaderB] = leaderA
                self.parentLevel[leaderA] += 1

class Edge:
    def __init__(self, a, b, w):
        self.a = a
        self.b = b
        self.w = w

def main():
    n, e = map(int, input().split())
    v = []
    ans = []
    dsu = DSU(n)
    for _ in range(e):
        a, b, w = map(int, input().split())
        v.append(Edge(a, b, w))
    v.sort(key=lambda x: x.w)
    cnt = 0
    cost = 0
    for val in v:
        a = val.a
        b = val.b
        w = val.w
        leaderA = dsu.find(a)
        leaderB = dsu.find(b)
        if leaderA == leaderB:
            cnt += 1
            continue
        ans.append(val)
        cost += val.w
        dsu.union(a, b)
    hs = all(dsu.find(i) == dsu.find(1) for i in range(1, n + 1))
    if hs:
        print(cnt, cost)
    else:
        print("Not Possible")

if __name__ == "__main__":
    main()

File: King_s_Order.java
import java.util.*;

class DSU {
    private long[] parent;
    private long[] parentLevel;
    private long[] parentSize;

    public DSU(int n) {
        parent = new long[n + 1];
        parentLevel = new long[n + 1];
        parentSize = new long[n + 1];
        Arrays.fill(parent, -1L);
    }

    public long find(long n) {
        if (parent[(int) n] == -1)
            return n;
        long leader = find(parent[(int) n]);
        parent[(int) n] = leader;
        return leader;
    }

    public void union(long a, long b) {
        long leaderA = find(a);
        long leaderB = find(b);
        if (leaderA != leaderB) {
            if (parentLevel[(int) leaderA] > parentLevel[(int) leaderB]) {
                parent[(int) leaderB] = leaderA;
            } else if (parentLevel[(int) leaderB] > parentLevel[(int) leaderA]) {
                parent[(int) leaderA] = leaderB;
            } else {
                parent[(int) leaderB] = leaderA;
                parentLevel[(int) leaderA]++;
            }
        }
    }
}

class Edge implements Comparable<Edge> {
    public long a, b, w;

    public Edge(long a, long b, long w) {
        this.a = a;
        this.b = b;
        this.w = w;
    }

    @Override
    public int compareTo(Edge other) {
        return Long.compare(this.w, other.w);
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int e = scanner.nextInt();
        List<Edge> v = new ArrayList<>();
        List<Edge> ans = new ArrayList<>();
        DSU dsu = new DSU(n);

        for (int i = 0; i < e; i++) {
            long a = scanner.nextLong();
            long b = scanner.nextLong();
            long w = scanner.nextLong();
            v.add(new Edge(a, b, w));
        }

        Collections.sort(v);

        int cnt = 0;
        long cost = 0;

        for (Edge val : v) {
            long a = val.a;
            long b = val.b;
            long w = val.w;
            long leaderA = dsu.find(a);
            long leaderB = dsu.find(b);

            if (leaderA == leaderB) {
                cnt++;
                continue;
            }

            ans.add(val);
            cost += val.w;
            dsu.union(a, b);
        }

        boolean hs = true;

        for (int i = 1; i <= n; i++) {
            if (dsu.find(i) != dsu.find(1)) {
                hs = false;
                break;
            }
        }

        if (hs) {
            System.out.println(cnt + " " + cost);
        } else {
            System.out.println("Not Possible");
        }
    }
}
